### DISCUSSION TITLE
Update method is not virtual

### DISCUSSION THREAD

**Discussion by michele-bettin** [CollaboratingPlatypus/PetaPoco#652] [Q&A] [answered]
Hi everyone, I have to create an **audit** table for all updated records in the database but I was forbidden to use sql procedure like triggers/controls/events... So, I decided to inherit the **Database** class of _Petapoco_ and override the **Update** method in order to store my audit record like everytime an **Update()** is call in my code. Every record in audit should track the object changed in json format, the type of the object changed, the user and the date. The problem here is that Upload method is _private_ and so, it is not overridable... Do you have some suggestions for me to solve this issue using only the back-end code and without using sql code?
Furthermore, I can't use and override the **OnExecutingCommand()** because I need the type of the object that I'm _updating_, so I'm stucked...

**Comment by asherber** [comment, marked as answer]
First, a slight correction: `Database.Update()` is public, not private, but you're right that it's not virtual and therefore can't be overridden.

However, you can create your own descendant class and re-implement the `Update()` methods with the `new` keyword. Many of the `Update()` overloads just call other overloads; it looks like there are 4 you'd have to re-implement. You'd also have to make sure that whenever you're passing around a database variable, the variable is of your own type rather than `Database` or `IDatabase` -- otherwise your methods won't get called.

Here's an example of what I mean.

```csharp
public class MyDatabase: Database
{
    // Need to implement some constructors and call the base constructor
    
    public new int Update(object poco)
    {
        // Do your audit logging here, then:
        return base.Update(poco);
    }
}
```
A second approach would be to create your own class that wraps `Database` instead of inheriting from it. The downside is that you'd need to explicitly implement any `Database` methods you want to use and forward them to the actual `Database`.

```csharp
public class MyDatabaseWrapper
{
    private readonly Database _database;

    public MyDatabaseWrapper()
    {
        // instantiate _database
    }

    public int Update(object poco)
    {
        // do your audit logging, then:
        return _database.Update(poco);
    }
}
```

A third option, possibly the simplest, would be to define some extension methods to do what you want.

```csharp
public static class DatabaseExtensions
{
    public static int AuditAndUpdate(this Database db, object poco)
    {
        // Do audit logging, then:
        return db.Update(poco);
    }
}

var db = new Database(...);
db.AuditAndUpdate(somePoco);
```

**Reply by michele-bettin** [reply]
@asherber Your answer is awesome, thank you so much!!!
